Build python from source for CI

aritra24 opened the pull request:
closes: #52106

Builds python from source, also installs
golang from official distribution. Does
both of these for the ci image only.

<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->



<!-- Please keep an empty line above the dashes. -->
---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [airflow-core/newsfragments](https://github.com/apache/airflow/tree/main/airflow-core/newsfragments).

aritra24 committed:
Draft: Build python from source

Builds python from source, also installs
golang from official distribution. Does
both of these for the ci image only.

aritra24 committed:
Updates path

potiuk commented on scripts/docker/install_os_dependencies_ci.sh:
```diff
@@ -0,0 +1,166 @@
+set -euo pipefail
+
+if [[ "$#" != 1 ]]; then
+    echo "ERROR! There should be 'runtime', 'ci' or 'dev' parameter passed as argument.".
```
```suggestion
    echo "ERROR! There should be 'RUNTIME', 'CI' or 'DEV' parameter passed as argument.".
```

potiuk replied:
ah no

potiuk reviewed (approved):
Looks great and is IMHO "fast enough" - enough to merge and try it even now in CI image

potiuk commented on scripts/docker/install_os_dependencies_ci.sh:
```diff
@@ -0,0 +1,166 @@
+}
+
+function install_python() {
+    git clone --branch ${PYTHON_MAJOR_MINOR_VERSION} --depth 1 https://github.com/python/cpython.git
```
One small caveat here. The way how it works is that it will cache that layer the first time the branch is checked out and we have no mechanism currently to invalidate the cache - which will mean that we will be stuck with the version of Python sources that were released at the moment we built it for the first time in CI.

What we really need to do is likely to:

1) specify the exact python version with patchlevel in the image somehow (different in each version) - in the way that it is different for each MAJOR/MINOR version

Something like (:

```
ARG PYTHON_VERSION=3.9.23
ENV PYTHON_VERSION=${PYTHON_VERSION}
```

This will invalidate the cache at the moment we change version


2) We should modify our pre-commit upgrade-check to check which versions of python are currently released and fail the check in the canary build when we have outdated version

aritra24 committed:
Adds version upgrade check for python version

Adds support for using the airflow api to fetch
the newest python patch version available for specific
major_minor pair

aritra24 committed:
Merge branch 'main' into install_python_from_source

potiuk commented on .github/workflows/basic-tests.yml:
```diff
@@ -292,6 +292,7 @@ jobs:
         if: always()
         env:
           UPGRADE_UV: "true"
+          UPGRADE_PYTHON: "false"
```
As discussed in slack it would also be great to add golang here - in the fashion similar to UV - we do not "fail" the build when uv has new version - we merely print the output showing that it **should** be upgraded, but when we run pre-commit locally we use defaults - which means we update everything. That is the best of both worlds.

potiuk reviewed (approved):
Really nice - with one caveat only - that we should add golang in a similar fashion as `uv`

aritra24 committed:
Updated to use args in dockerfile for python

aritra24 committed:
Added support for golang upgrade

aritra24 committed:
Fixed go version sorting in pre_commit install

aritra24 committed:
Merge branch 'main' into install_python_from_source

gopidesupavan commented on scripts/ci/pre_commit/update_installers_and_pre_commit.py:
```diff
@@ -65,6 +66,31 @@ def get_latest_pypi_version(package_name: str) -> str:
+    latest_version = None
+    version_match = re.compile(rf"^v{python_major_minor}.*\.\d+$")
+    for i in range(5):
+        response = requests.get(
```
If possible lets use GITHUB_TOKEN here please.

potiuk replied:
good point

aritra24 replied:
Ack, will do

aritra24 replied:
done

gopidesupavan commented on scripts/ci/pre_commit/update_installers_and_pre_commit.py:
```diff
@@ -65,6 +66,31 @@ def get_latest_pypi_version(package_name: str) -> str:
+        data = response.json()
+        versions = [str(tag["name"]) for tag in data if version_match.match(tag.get("name", ""))]
+        if versions:
+            latest_version = sorted(versions, key=Version, reverse=True)[0]
```
This throws error if any version tag is not in the expected format of `Packaging version`, probably removing bad tags and checking would be better?

aritra24 replied:
We're already filtering here for versions https://github.com/apache/airflow/pull/52265/files/f00fbf0985a906f29af78602b034bccb4cd4b757#diff-58ab88b772db4e1d35af29ef3e1220966641a855181c8629d5ba77c73196fd75R71

potiuk replied:
But yeah. I think it will be safer to simply discard those that are not parseable by Version. Never trust a regexp.

aritra24 replied:
They're also using a regex internally, just a more complex one 🙃 
```
    v?
    (?:
        (?:(?P<epoch>[0-9]+)!)?                           # epoch
        (?P<release>[0-9]+(?:\.[0-9]+)*)                  # release segment
        (?P<pre>                                          # pre-release
            [-_\.]?
            (?P<pre_l>alpha|a|beta|b|preview|pre|c|rc)
            [-_\.]?
            (?P<pre_n>[0-9]+)?
        )?
        (?P<post>                                         # post release
            (?:-(?P<post_n1>[0-9]+))
            |
            (?:
                [-_\.]?
                (?P<post_l>post|rev|r)
                [-_\.]?
                (?P<post_n2>[0-9]+)?
            )
        )?
        (?P<dev>                                          # dev release
            [-_\.]?
            (?P<dev_l>dev)
            [-_\.]?
            (?P<dev_n>[0-9]+)?
        )?
    )
    (?:\+(?P<local>[a-z0-9]+(?:[-_\.][a-z0-9]+)*))?       # local version
```

aritra24 replied:
they don't seem to have a separate is valid method, I chose to do it with regex myself because I wanted to exclude alpha and beta versions from our list which would've been matched by their regex, so I think given our regex is a subset of theirs it should be fine?

potiuk replied:
> They're also using a regex internally, just a more complex one 🙃 

He he. But it's THEM to manage it .... :). and they're canonical. Also it's a bit far from the original filter to that place. How about:

```python
def is_valid_version(v):
    try:
        Version(v)
        return True
    except InvalidVersion:
        return False

valid_versions = list(filter(is_valid_version, versions))
```

potiuk replied:
Also. We can do it better without earlier filtering:

```python
def is_valid_final_version(v):
    try:
       version = Version(v)
       return version == version.base_version
    except InvalidVersion:
        return False

valid_final_versions = list(filter(is_valid_final_version, versions))

````

potiuk replied:
(we still have to remove v possibly)

potiuk replied:
You know .. there is this saying..

> You have problem, introduce regexp, now you have two problems.

potiuk replied:
Of course all those are NITs... It can go without it :)

potiuk replied:
It's just a bit more explicit and canonical

gopidesupavan reviewed (approved):
@aritra24 Great work :) just one small nit everything fine. Thanks for doing this.

aritra24 committed:
Added github token usage and fixed version regex

Updated python fetch request during upgrade to use
github token and fixed the regex

aritra24 commented on Dockerfile.ci:
```diff
@@ -1279,8 +1308,14 @@ ENV DEV_APT_COMMAND=${DEV_APT_COMMAND} \
+RUN AIRFLOW_PYTHON_VERSION=$AIRFLOW_PYTHON_VERSION \
+    GOLANG_MAJOR_MINOR_VERSION=${GOLANG_MAJOR_MINOR_VERSION} \
+    bash /scripts/docker/install_os_dependencies_ci.sh ci
 
```
One minor note here, I had to pass the args explicitly into the run command because without them the layers weren't getting invalidated since docker tries to be smart that way to reduce unnecessary builds. 
cc: @potiuk

potiuk replied:
Define them as ENV before that will invalidate the layers after.

potiuk replied:
This is what we do in other places.

aritra24 replied:
I think I did try that, it still was not invalidating the cache 🤔 I can recheck once more though

potiuk replied:
<img width="775" alt="Screenshot 2025-07-05 at 08 43 38" src="https://github.com/user-attachments/assets/79976bff-4fb8-45bb-886f-00408b8de15d" />

potiuk replied:
*must work* if you pass the value from ARG .. The thing is it will **not** invalidate layer when you already built it before with that ARG and set that ENV VAR before -- it will keep the built layer in cache and see that it already built it.

aritra24 replied:
Yeah, I think that was it, I cleaned up my build cache and then it did rebuild it

aritra24 replied:
works now

aritra24 committed:
Updated dockerfile.ci file

aritra24 committed:
Merge branch 'main' into install_python_from_source
